### PR TITLE
add '-Dusethreads' to perl build Configure

### PR DESCRIPTION
**Pull request by tomkinsc**
Closes https://github.com/conda-forge/perl-feedstock/pull/5
Fixes https://github.com/conda-forge/perl-feedstock/issues/4 ?

The synchronization overhead of perl with threading support is minimal. Single-threaded perl programs will not see a significant performance penalty from being run on perl supporting threads (at most [about 5%](https://mail-archives.apache.org/mod_mbox/perl-modperl/200311.mbox/%3C3FB298A0.3050503@stason.org%3E)), however without threading support, multi-threaded programs cannot run.
It would be a sane default to include ‘-Dusethreads’ in the perl build script, so any packages that depend on ‘perl’ will run, whether they are single- or multi-threaded. It may make sense to add a second feedstock for something like ‘perl-singlethreaded’ for those who need the absolute best efficiency for their single-threaded perl applications.

This includes the flag ordering change present in https://github.com/conda-forge/perl-feedstock/pull/5 where `-de` is added at the end, since apparently the order matters.

Edit by @jakirkham : Added closes/fixes lines above.

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-sh Configure -de -Dprefix=$PREFIX -Duserelocatableinc
+sh Configure -Dusethreads -Duserelocatableinc -Dprefix=$PREFIX -de
```
Comment (jakirkham) [inline recipe/build.sh]: So I believe there was something special about how these needed to be ordered. See issue ( https://github.com/conda-forge/perl-feedstock/issues/4 ). If you could explain a bit why this order is better, that would be very helpful.

Comment (jakirkham) [inline recipe/build.sh]: Also please see PR ( https://github.com/conda-forge/perl-feedstock/pull/5 ) where I was trying to do the same thing.

Comment (tomkinsc) [inline recipe/build.sh]: Yes, I placed `-de` at the end for that reason (saw your pull request).

**Comment by tomkinsc** [comment]
It looks like the build failed with errors such as this one:

```Running Makefile.PL in cpan/Pod-Checker
../../miniperl -I../../lib Makefile.PL INSTALLDIRS=perl INSTALLMAN1DIR=none INSTALLMAN3DIR=none PERL_CORE=1 LIBPERL_A=libperl.a
WARNING: Setting ABSTRACT via file 'lib/Pod/Checker.pm' failed```

Any ideas how to correct it?

**Comment by bgruening** [comment]
@tomkinsc no idea from my side :(

**Comment by bgruening** [comment]
@tomkinsc maybe the trick mentioned at the end of the error report helps?
```
### You may have to set your dynamic library search path,
### LD_LIBRARY_PATH, to point to the build directory:
###   setenv LD_LIBRARY_PATH `pwd`; cd t; ./perl harness
###   LD_LIBRARY_PATH=`pwd`; export LD_LIBRARY_PATH; cd t; ./perl harness
###   export LD_LIBRARY_PATH=`pwd`; cd t; ./perl harness
### for csh-style shells, like tcsh; or for traditional/modern
### Bourne-style shells, like bash, ksh, and zsh, respectively.
```

**Comment by bgruening** [comment]
Moreover, this `Insecure directory in $ENV{PATH} while running with -T switch at ../../t/test.pl line 738.` seems to indicate that a few files and folders are world-writable, which is not allowed.

**Comment by jakirkham** [comment]
If there are few Bioconda people interested in helping maintain Perl, that would be great. I think you guys use Perl more often than I or others do. So you probably have a better idea of how best to get it to work.

**Comment by bgruening** [comment]
@jakirkham we are here to help :)
@tomkinsc are the travis changes really needed?

**Comment by bgruening** [comment]
@tomkinsc I cancelled the previous builds so that the last one is faster :)

**Comment by tomkinsc** [comment]
@bgruening: Yes, the changes were needed since I had to update the CircleCI-related files, and Travis had a shared dependency. Thanks for cancelling the previous builds! :)

**Comment by tomkinsc** [comment]
Not sure about the AppVeyor failure.

**Comment by bgruening** [comment]
It seems that the appveyor config also needs some updates? Other feedstocks have a different config.
@jakirkham I thought conda-smithy is updating these configs, no?

**Comment by jakirkham** [comment]
So we have a webservice that re-renders a random subset of feedstocks daily. However, the number of feedstocks has really grown and that approach is proving itself insufficient for this scale. We have been discussing an asynchronous webservice that PRs re-renderings against feedstocks/PRs when they get updated. ( https://github.com/conda-forge/conda-forge-webservices/issues/74 ) That should do a better job going forward.

In the near term, I have put a fair bit of effort into just making `conda-smithy` easier to use and [documenting]( https://github.com/conda-forge/conda-smithy#re-rendering-an-existing-feedstock ) it better in the hopes that people can do this themselves as needed.

**Comment by bgruening** [comment]
@jakirkham I guess this is fair enough. We should just know that smithy is not taking care of it :)

**Comment by jakirkham** [comment]
Also the Windows build didn't/doesn't actually build Perl. Instead it installed a pre-built binary of Perl. This is what @msarahan found to work at the time. Hence why it didn't/shouldn't have a Python matrix on AppVeyor.

**Comment by bgruening** [comment]
So `skip: True  # [win]` is the way forward? Good to know.

**Comment by jakirkham** [comment]
Uh, sorry, I think the wires got crossed. What I mean is there is a Windows build, but it just copy StrawberryPerl over into the conda build prefix. We just don't need the typical `python` build dependency or `vc` features.

**Comment by tomkinsc** [comment]
I don't have time to address this, so I'm closing this PR; feel free to fork my branch. Setting up various tokens and "re-rendering" the automation infrastructure is a silly amount of overhead when I simply wanted to add a flag to the conda build script.

**Comment by jakirkham** [comment]
Sorry you are having a hard time of it.

So there shouldn't be any tokens that one needs to set.

I can open a re-rendering PR that you can base this on top of if you would like.

**Comment by jakirkham** [comment]
I've added PR ( https://github.com/conda-forge/perl-feedstock/pull/7 ), which will re-render the feedstock. Given this is a very old feedstock, let's wait and see how that goes first. If it goes ok, we can continue with these adjustments.

**Comment by jakirkham** [comment]
FYI the re-rendering has been merged.

**Comment by tomkinsc** [comment]
Thanks for re-rendering the feedstock. I've merged the updated master into this branch.

**Comment by jakirkham** [comment]
Of course, glad to help.

It seems there are some subtle differences on some of the CI files. Could you please click the [Allow edits from maintainers]( https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ ) box and I can push a fix to clean those up directly to your PR?

**Comment by tomkinsc** [comment]
Done.

**Comment by jakirkham** [comment]
Thanks.

File: recipe/build.sh [hunk 2]
```diff
@@ -1,5 +1,23 @@
 #!/bin/bash
+
+if [ -z "$LD_LIBRARY_PATH" ]; then
+    old_ld_library_path="$LD_LIBRARY_PATH"
+fi
+
+export LD_LIBRARY_PATH=$(pwd)
```
Comment (jakirkham) [inline recipe/build.sh]: Could you please explain why this change was needed?

File: recipe/build.sh [hunk 3]
```diff
@@ -1,5 +1,23 @@
...
+
+# world-writable files are not allowed
+chmod -R o-w $SRC_DIR
+
 sh Configure -de -Dprefix=$PREFIX -Duserelocatableinc
 make
+
+# change permissions again after building
+chmod -R o-w $SRC_DIR
```
Comment (jakirkham) [inline recipe/build.sh]: ...and how they become wrong again.

File: recipe/build.sh [hunk 4]
```diff
@@ -1,5 +1,23 @@
...
+if [ -z "$LD_LIBRARY_PATH" ]; then
+    old_ld_library_path="$LD_LIBRARY_PATH"
+fi
+
+export LD_LIBRARY_PATH=$(pwd)
+
+# world-writable files are not allowed
+chmod -R o-w $SRC_DIR
+
 sh Configure -de -Dprefix=$PREFIX -Duserelocatableinc
```
Comment (jakirkham) [inline recipe/build.sh]: Are we still planning to add `-Dusethreads`?

File: recipe/build.sh [hunk 5]
```diff
@@ -1,5 +1,23 @@
...
+
+# change permissions again after building
+chmod -R o-w $SRC_DIR
+
 make test
 make install
+
+if [ ! -z "$old_ld_library_path" ]; then
+    export LD_LIBRARY_PATH="$old_ld_library_path"
+fi
```
Comment (jakirkham) [inline recipe/build.sh]: Why do we need to set it back? Doesn't `conda-build` override this anyways?

File: recipe/build.sh [hunk 6]
```diff
@@ -1,5 +1,23 @@
 #!/bin/bash
+
+if [ -z "$LD_LIBRARY_PATH" ]; then
+    old_ld_library_path="$LD_LIBRARY_PATH"
+fi
+
+export LD_LIBRARY_PATH=$(pwd)
+
+# world-writable files are not allowed
+chmod -R o-w $SRC_DIR
```
Comment (tomkinsc) [inline recipe/build.sh]: Incorrect as built by perl?

File: recipe/build.sh (hunk 3 above)
Comment (tomkinsc) [inline recipe/build.sh]: Built files may have incorrect permissions, so the script corrects before building and again after. Tests were previously failing because various files were world-writable.

File: recipe/build.sh (hunk 4 above)
Comment (tomkinsc) [inline recipe/build.sh]: Not sure how it was lost in this PR, but yes.

File: recipe/build.sh (hunk 5 above)
Comment (tomkinsc) [inline recipe/build.sh]: If conda-build will restore it, then this can be removed.

File: recipe/build.sh (hunk 4 above)
Comment (bgruening) [inline recipe/build.sh]: Ok, I guess this is needed before we can ask for a merge :)

File: recipe/build.sh (hunk 5 above)
Comment (bgruening) [inline recipe/build.sh]: I agree, this can be removed.

**Comment by bgruening** [comment]
@tomkinsc do you have time to finishe this one?

**Comment by bgruening** [comment]
@jakirkham ready to be merged? :)
Thanks a lot @tomkinsc!

**Comment by jjhelmus** [comment]
Any idea of these changes would make perl relocatable (ie fix #4)?  I've been trying to re-build the conda-forge `autoconf` package with `conda-build` `2.x` and something with the current perl build is causing issues.

I'll move this topic to it's own issue this is unrelated to the current changes.

**Comment by jjhelmus** [comment]
Please ignore my last comment.  The failure I was seeing is unrelated to the perl package.  Sorry for any confusion.

**Comment by tomkinsc** [comment]
Ready to merge!

File: recipe/build.sh [hunk 7]
```diff
@@ -1,5 +1,15 @@
...
+
+sh Configure -Dusethreads -Duserelocatableinc -Dprefix=$PREFIX -de
 make
+
+# change permissions again after building
+chmod -R o-w $SRC_DIR
+
 make test
-make install
+make install
```
Comment (jakirkham) [inline recipe/build.sh]: Note: Fixed to a terminal EOF again.
